src/Microsoft.OData.Client/DataServiceContext.cs

```diff
@@ -3935,13 +4018,89 @@ private void DeleteObjectInternal(object entity, bool failIfInAddedState)
                 // or Modified state referencing a source/target entity in the Deleted state.
                 resource.State = EntityStates.Deleted;
+                resource.ChangeSetId = Guid.NewGuid().ToString();
                 this.entityTracker.IncrementChange(resource);
             }
         }
 
+        /// <summary>
+        /// Mark an existing object being tracked by the context for deletion.
+        /// </summary>
+        /// <param name="entity">entity to be mark deleted</param>
+        /// <param name="failIfInAddedState">If true, then an exception will be thrown if the entity is in the added state.</param>
+        /// <exception cref="ArgumentNullException">if entity is null</exception>
+        /// <exception cref="InvalidOperationException">if entity is not being tracked by the context, or if the entity is in the added state and <paramref name="failIfInAddedState"/> is true.</exception>
+        /// <remarks>
+        /// Existing objects in the Added state become detached if <paramref name="failIfInAddedState"/> is false.
+        /// </remarks>
+        private void DeleteObjectInternal(object entity, bool failIfInAddedState)
+        {
+            this.DeleteObjectInternal(entity, failIfInAddedState, null);
+        }
+
+        private static void GetDependsOnChangeOrdersAndChangeSetIds(
+            List<string> dependsOnIds,
+            IEnumerable<EntityDescriptor> entities,
+            ClientEdmModel model,
+            out List<string>  dependsOnIdsAsChangeOrders,
+            out HashSet<string> dependsOnChangeSetIds)
+        {
+            dependsOnIdsAsChangeOrders = new List<string>();
+            dependsOnChangeSetIds = new HashSet<string>();
+
+            foreach (string dependOnId in dependsOnIds)
+            {
+                foreach (EntityDescriptor entityDescriptor in entities)
+                {
+                    object entity = (object)entityDescriptor.Entity;
+                    string entityName = entity.GetType().FullName;
+                    string key = GetKeyPropertyName(model, entityName);
+                    object id = entity.GetType().GetProperty(key).GetValue(entity, null);
+                    if (dependOnId == id.ToString())
+                    {
+                        uint changeOrder = entityDescriptor.ChangeOrder;
+                        dependsOnIdsAsChangeOrders.Add(changeOrder.ToString(CultureInfo.InvariantCulture));
+                        dependsOnChangeSetIds.Add(entityDescriptor.ChangeSetId);
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Returns the name of the Declared Key for an entity.
+        /// </summary>
+        /// <param name="model"></param>
+        /// <param name="entityName"></param>
+        /// <returns>Declared key property name.</returns>
+        private static string GetKeyPropertyName(ClientEdmModel model, string entityName)

```

Comment:
@KenitoInc They way you're going about identifying the key properties here seems very direct and I have to wonder about the logic in [`GetKeyPropertiesOnType` method](https://github.com/OData/odata.net/blob/325e28394a473e579ac7573e0928ebb3635b5765/src/Microsoft.OData.Client/Metadata/ClientTypeUtil.cs#L428). Would the existing method do the job efficiently?

Reply:
I updated the design of this feature.
We now have a straight forward way of getting the `ChangeOrder` without relying on the key property.